pumpkin-solver/src/propagators/gcc/gcc_extended_resolution/upper_bound.rs

```diff
@@ -88,54 +88,91 @@ impl<Var: IntegerVariable> Propagator for GccUpperBound<Var> {
             }
         }
 
-        dbg!(&uf);
-
-        let mut assigned: HashMap<i32, Vec<&Var>> = HashMap::default();
-        for var in &self.variables {
-            if context.is_fixed(var) {
-                let value = context.lower_bound(var);
-                assigned.entry(value).or_default().push(var);
-            }
-        }
+        // Now recalculated per set, to include only variables that aren't in the set
+        // See `assigned_not_in_set`
+        // let mut assigned: HashMap<i32, Vec<&Var>> = HashMap::default();
+        // for var in &self.variables {
+        //     if context.is_fixed(var) {
+        //         let value = context.lower_bound(var);
+        //         assigned.entry(value).or_default().push(var);
+        //     }
+        // }
 
         for set in uf.subsets() {
             if set.len() < 2 {
                 // Ignore sets with only one variable
                 continue;
             }
 
+            // Iterating through HashSet isn't stable, so copy into a Vec
+            let set_vec: Vec<usize> = set.iter().cloned().collect();
+            let mut set_reason = Vec::new();
+            for (i, &elem_1) in set_vec.iter().enumerate() {
+                for &elem_2 in set_vec.iter().skip(i + 1) {
+                    let lit = self.get_equality(elem_1, elem_2);
+                    if context.is_literal_true(&lit) {
+                        set_reason.push(predicate!(lit == 1));
+                    }
```

Comment:
I think this still has the problem of introducing non-propagating reasons. So if we have x=y, y=z, x=z, then we should only include 2 of them. Or, in general, we should include a cycle-free explanation, i.e. a tree.